clearnode/ledger_transaction.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/shopspring/decimal"
+	"gorm.io/gorm"
+)
+
+type TransactionType int
+
+const (
+	TransactionTypeTransfer      TransactionType = 100
+	TransactionTypeDeposit       TransactionType = 201
+	TransactionTypeWithdrawal    TransactionType = 202
+	TransactionTypeAppDeposit    TransactionType = 301
+	TransactionTypeAppWithdrawal TransactionType = 302
+)
+
+var (
+	ErrInvalidLedgerTransactionType = errors.New("invalid ledger transaction type")
+)
+
+type LedgerTransaction struct {
+	ID          uint            `gorm:"primaryKey"`
+	Hash        string          `gorm:"column:hash;not null;uniqueIndex"`
```

Comment:
I love the idea of having transaction hashes, but at this point it brings a lot of complexity ( Before and After create gorm hooks). I don't want to rely on this type of behaviour, as for me transaction hash should be calculated without depending on DB state ( id and createdAt). I would prefer if they were deterministic.

At the same time I think that we can spend a lof of time designing this properly, so let's drop it for now and get back to this topic later, when we will be able to involve more people.